util/stream_request.py

```diff
@@ -193,11 +209,129 @@ def insert_provenance(self):
+            # only propagate failures on affected parameters whose upstream parameters are not inputs to this stream
+            for matched_parameter in failed_nearby_parameters.difference(stream_parameters):
+                propagate_input = failed_nearby_params_map[matched_parameter]
+                qc_variances = propagate_input[0]
+                affected_parameters = propagate_input[1]
+                for propagate_affected in affected_parameters.intersection(stream_parameters):
+                    param_qc_executed = propagate_affected.name + QC_EXECUTED
+                    for dataset in stream_dataset.datasets.itervalues():
+                        if param_qc_executed in dataset.keys():
+                            qc_executed_values = dataset[param_qc_executed].values
+                            for position, variance in qc_variances:
+                                # propagate the failure on this particle if propagating tests failed
+                                failed_tests_to_propagate = [(i, self.qc_test_flag_map[i][0])
+                                                             for i in self.qc_test_flag_map.keys()
+                                                             if i & variance and self.qc_test_flag_map[i][1]]
+                                if failed_tests_to_propagate:
+                                    qc_executed_values[position] = \
+                                            qc_executed_values[position] | self.qc_propagate_flag
+
+            failed_related_parameters = propagate_failures.get((RELATED, sk.as_three_part_refdes(), sk.stream_name,
+                                                                sk.method), {})
+            for propagate_input in failed_related_parameters.values():
+                qc_variances = propagate_input[0]
+                affected_parameters = propagate_input[1]
+                for propagate_affected in affected_parameters.intersection(stream_parameters):
+                    param_qc_executed = propagate_affected.name + QC_EXECUTED
+                    for dataset in stream_dataset.datasets.itervalues():
+                        if param_qc_executed in dataset.keys():
+                            qc_executed_values = dataset[param_qc_executed].values
+                            for position, variance in qc_variances:
+                                # propagate the failure on this particle if propagating tests failed
+                                failed_tests_to_propagate = [(i, self.qc_test_flag_map[i][0])
+                                                             for i in self.qc_test_flag_map.keys()
+                                                             if i & variance and self.qc_test_flag_map[i][1]]
+                                if failed_tests_to_propagate:
+                                    qc_executed_values[position] = \
+                                            qc_executed_values[position] | self.qc_propagate_flag
```

Comment:
I see that comments 12 and 13 of the 12189 ticket mention this, but I don't understand why QC is propagated by changing bits in the qc_executed parameter instead of the qc_results parameter. My understanding was that <param>_qc_executed indicated what QC tests were run for param and <param>_qc_results indicated which QC tests passed for param. It seems strange to me, therefore, to indicate a propagated failure in qc_executed.

Reply:
Nevermind, this should be OK. I wasn't realizing there is a special qc propagation flag and thought other test flags were getting overwritten.